<html>
<head>
<link type="text/css" media="all" rel="stylesheet" href="/gbrowse2/css/gbrowse.css" />
<link type="text/css" media="all" rel="stylesheet" href="/gbrowse2/css/gbrowse_warm.css" />
</head>
<body>

<table width="100%"  style="border:ridge">
  <tr class="searchtitle">
    <th><a href="#format">File Formats</a></th>
    <th><a href="#bedfile format">BED File Format</a></th>
    <th><a href="#featurefile format">Feature File Format</a></th>
    <th><a href="#customize">Customizing Appearance</a></th>
    <th><a href="#numeric">Displaying Intensity Plots &amp; Other Numeric Data</a></th>
    <th><a href="#remote">Importing Custom Tracks via the Internet</a></th>
  </tr>
</table>

<A NAME="intro"></A>
<h1>Uploading Custom Tracks</h1>

<p>

This genome browser allows you to upload custom track data in a
variety of quantitative and non-quantitative formats. You can also <a
href="#remote">import data sets</a> that reside on network-connected
servers by referring to their URLs.

<p>

Accepted formats include:

<ul>
  <li><a href="http://genome.ucsc.edu/FAQ/FAQformat#format1"
  target="_blank">BED</a>

  <li><a href="#featurefile format"  target="_blank">GBrowse Feature
  File Format</a>

  <li> <a href="http://genome.ucsc.edu/FAQ/FAQformat#format3" target="_blank">GFF</a>

  <li><a href="http://www.sequenceontology.org/gff3.shtml" target="_blank">GFF3</a>

  <li> <a href="http://genome.ucsc.edu/goldenPath/help/wiggle.html" target="_blank">Wiggle (WIG)</a> (for dense quantitative data)

  <li><a href="http://samtools.sourceforge.net/"
target="_blank">BAM or SAM</a> (for sequence alignment data)
</ul>

<p>

Please see the links above for descriptions of how to create and
format these data files.

<p>

In addition, GBrowse supports an internal genome annotation file
format that is unimaginatively called "Feature File Format" (FFF). It
is a relatively simple format that allows you to create features, name
them, group them together, and tweak their display in various ways.

<p>

Both the FFF and GFF3 formats allow you to customize the display using
GBrowse's track configuration sections, which are described later in
this document. When uploading WIG or BED files, GBrowse will convert
the UCSC track definition line into an equivalent GBrowse
configuration section.

<a name="bedfile format"></a>

<h2>BED Format</h2>

<p>
The simplest format to use is BED. Each line of the file is a feature; each feature
has three required fields, and 9 optional fields, separated by white space:

<ol>
  <li><b>chrom</b> - The name of the chromosome (e.g. chr1), contig or scaffold.
  <li><b>chromStart</b> - The starting position of the feature on the chromosome. The first base of the chromosome is <b>numbered 0</b>.
  <li><b>chromEnd</b> - The ending position of the feature on the chromosome. The chromEnd base is not part of the feature.
</ol>

<p>

For example, this file will define a single 100 bp feature on chromosome chr1 spanning the bases 0 to 99:
<blockquote>
<pre>
chr1 0 100
</pre>
</blockquote>

<p>

Additional fields let you add a name, a score, a color and other types of decoration to the feature. The most commonly
used optional field is the name field, which allows you to add a name
to your feature:

<blockquote>
<pre>
chr1 0 100 myFavoriteGene
</pre>
</blockquote>


<p>

Please 
see the <a href="http://genome.ucsc.edu/FAQ/FAQformat#format1">UCSC genome browser documentation</a> for 
the full description of everything you can do with BED. 

<a name="featurefile format"></a>

<h2>Feature File Format</h2>

<p>

In addition to the BED file format for qualitative tracks, GBrowse
supports "Feature File format", which is slightly more 
configurable than BED. In addition, it uses 1-based chromosome
coordinates, which are somewhat more intuitive.

<p>

Here is an example:

<blockquote>
<pre>
[mRNA]
glyph = gene
key   = My Genes

mRNA hox1 Chr1:1..100 Type=UTR
mRNA hox1 Chr1:101..200,300..400,500..800 Type=CDS
mRNA hox1 Chr1:801..1000 Type=UTR

[EST]
glyph   = segments
bgcolor = green
key     = Mapped Expressed Tags

EST	yk260e10.5	Chr3:15569..15724
EST	yk672a12.5	Chr3:537..618,3187..3294
EST	yk595e6.5	Chr3:552..618

[PRED]
glyph       = heat_map
start_color = yellow
end_color   = blue
min_score   = 0
max_score   = 100
citation    = Bluer exons have higher prediction values
key         = Predicted genes

PRED	"Gene 1"	Chr1:518..616 score=50
PRED    "Gene 1"        Chr1:661..735 score=80
PRED    "Gene 1"        Chr1:3187..3365 score=50
PRED	"Gene 2"	Chr1:16626..17396 score=85
PRED    "Gene 2"        Chr1:17451..17597 score=90
</pre>
</blockquote>

<h3>The Data Lines</h3>

<p>

Each annotation occupies one or more lines.  It contains three to five
columns, delimited by tabs or spaces:

<dl>
  <dt><b>Column 1, the feature type</b>
  <dd>The first column is the feature type.  Any description is valid,
      but a short word, like "knockout" is better than a long one, like
      "Transposon-mediated knockout".  Later on you can provide a long descriptive
      name in the formatting key if you desire.  If the feature type contains white
      space, you must surround it by double or single quotes.
  <dt><b>Column 2, the feature name</b>
  <dd>This is the name of the feature.  The name will be displayed
      above the feature when there's room for it and name display is turned on.
      Shorter names are more attractive than long ones.  If the name contains white space, you must
      surround it by white space.  Use empty quotes ("") if there is no name to display.
  <dt><b>Column 3, the feature position</b>
  <dd>The third column contains one or more ranges occupied by the
      feature.  A range has a sequence ID indicating the chromosome, contig or other reference sequence
      on which the feature resides, plus a start and end position, and is expressed either as
      "seqID:start..stop" or "seqID:start-stop".  Use whichever form
      you prefer. You can express a feature that occupies a discontinuous set of ranges, such as an mRNA
      aligned to the genome, by providing a list of ranges separated by commas.  Example:
      <blockquote><pre>
      chr3:1..10,49..80,110..200
      </pre></blockquote>
      There should be no spaces before or after the commas.  If there are, enclose the entire range
      in quotes.
      <p>
      To describe an oriented feature that is on the minus strand, such as a transcribed gene,
      simply reverse the order of the start and stop ranges.  For example:
      <blockquote><pre>
      chr1:200..110,80..49,10..1
      </pre></blockquote>
      The strandedness is only displayed when using an arrowhead glyph, such the "transcript" glyph
      or the generic glyph  with the <b>strand_arrow=1</b> option.  See <a href="#formatting">Customizing the display</a>
      for details.
      <p>
      All ranges uses the coordinate system of the most recently declared reference landmark.
  <dt><b>Column 4, Tags</b> [optional]
  <dd>The fourth column, if present, is treated as a description.  The description will be printed at the
      bottom of the feature.  If there is no description, either leave blank, or use empty quotes.  If there
      is whitespace in the description, surround it with quotes.  You can also enter <i>attribute=value</i>
      information here for processing by certain glyphs.  The combinations you are likely to use currently
      are <i>URL=http://some.place</i> to indicate a URL to link to, <i>Note="some note"</i> to provide a descriptive
      caption to print under the feature, <i>Score=XXXXX</i> (where <i>XXXX</i> is some numeric value)
      to give the feature a score for those glyphs that chart numeric values, and <i>Type="some type"</i>
      to create genes and other complex multipart features.
      <p>
      See
      <a href="#customize">Customizing the display</a> for details.
</dl>

<p>

If most of your features are on the same chromosome, then you can
insert a <i>reference=&lt;seqid&gt;</i> line to indicate the default
sequence ID:

<blockquote>
<pre>
reference=Chr1
PRED	"Gene 1"	518..616 score=50
PRED    "Gene 1"        661..735 score=80
PRED    "Gene 1"        3187..3365 score=50
PRED	"Gene 2"	16626..17396 score=85
PRED    "Gene 2"        17451..17597 score=90
</pre>
</blockquote>

<p>

Chr1 will now be the default chromosome until the next
<i>reference=</i> line is encountered.

<p>

In addition to this format, you may use the standard GFF format for
your data.  Details can be found <a
href="http://www.sanger.ac.uk/Users/rd/gff.html">at the Sanger
Centre</a>.

<h3>Features and Subparts</h3>

<p>

To create a feature that has multiple subparts, you can indicate the
type of each subpart using the <i>Type="some type"</i> tag. Usually
you will use this for coding gene transcripts when you want to
distinguish the coding and non-coding portions of the gene.

<p>

Here is a simple example:

<blockquote>
<pre>
mRNA B0511.1   Chr1:1..100                     Type=UTR;Note="Putative primase"
mRNA B0511.1   Chr1:101..200,300..400,500..800 Type=CDS
mRNA B0511.1   Chr1:801..1000                  Type=UTR
</pre>

The top level feature's primary tag will be "mRNA", as indicated in the first column.
It will contain five subparts, a 5' UTR spanning positions 1..100, a series of three
CDS (coding) regions, and a 3' UTR extending from positions 801..1000.

<p>

Additional tags that are placed in the first line of the feature, such
as the Note, will be applied to the top level. In this example, the
note "Putative primase" will be applied to the mRNA at the top level
of the feature:

<h3>Grouping</h3>

<p>

You can group related features together.  The layout will attempt to
keep grouped features together, and will connect them with a dotted or
solid line if the <b>connector</b> option is specified.

<p>

A group is created using a line that contains just two columns
consisting of the feature type and name.  This is followed by a series
of data lines in which the feature type is blank.  For example:

<blockquote>
<PRE>
cDNA-clone  yk53c10
            yk53c10.5       18892-19154
            yk53c10.3       15000-15500,15700-15800

[cDNA-clone]
glyph = segments
connector = dashed
</PRE>
</blockquote>

<P>

This example creates a group of type "cDNA-clone" named Yk53c10.  It
consists of two sub-features, one the 5' EST and the other the 3'
EST.  The two configuration section that follows this group says to
use the "segments" glyph and to connect the parts using a dashed
line.  This is described in more detail later.

<p>

You can add URLs and descriptions to the components of a group, but
not to the group as a whole.


<h3>Comments</h3>

<p>

You can place a comment in the annotation file by preceding it with a
pound sign (#).  Everything following the pound sign is ignored:

<blockquote><pre>
# this is a comment
</pre></blockquote>

<hr>

<A NAME="customize"></a>
<H2>Customizing the Display</H2>

<P>

The browser will generate a reasonable display of your annotations by
default.  However, when using either FFF or GFF3 formats, you can
customize the appearance extensively by including one or more
configuration sections in the annotation file.  In addition to
changing the size, color and shape of the graphical elements, you can
attach URLs to them so that the user will be taken to a web page of
your choice when he clicks on the feature.  <p>

Here is an example configuration section.  It can appear at the top of
xthe file, at the bottom, or interspersed among data sections:

<blockquote>
<pre>
 # example file
 [general]
 height = 12
 strand_arrow = 1

 [EST]
 glyph = segments
 bgcolor= yellow
 connector = dashed
 height = 5

 [FGENES]
 glyph = gene
 bgcolor = green
 description = 1
</PRE>
</blockquote>

<P>

The configuration section is divided into a set of sections, each one
labeled with a <b>[section title]</b>. The <b>[general]</b> section
specifies global options for the entire image. Other sections apply to
particular feature types. In the example, the configuration in the
<b>[EST]</b> section applies to features labeled as having type "EST",
while the configuration in the <b>[FGENES]</b> section applies to
features labeled as predictions from the FGENES gene prediction
program. Options in more specific sections override those in the
general section.</P>

<P>Inside each section is a series of <b><i>name</i>=<i>value</i></b>
pairs, where the name is the name of an option to set. You can put
whitespace around the = sign to make it more readable, or even use a
colon (:) instead if you prefer. The following option names are
recognized:</P>


<table border="1">
  <tr class="searchtitle">
     <th>Option</th><th>Value</th><th>Example</th>
  </tr>

   <tr class="searchbody">
   <th>bgcolor</th>
   <td>Background color of each element</td>
   <td>blue</td>
 </tr>

  <tr class="searchbody">
    <th>bump</th>
    <td>Prevent features from colliding (0=no, 1=yes)</td>
    <td>1</td>
  </tr>

  <tr class="searchbody">
     <th>connector</th>
     <td>Type of group connector (dashed, hat or solid)</td>
     <td>dashed</td>
  </tr>

  <tr class="searchbody">
    <th>description</th>
    <td>Whether to print the feature's description (0=no, 1=yes)
    <td>0</td>
  </tr>

  <tr class="searchbody">
   <th>fgcolor</th>
   <td>Foreground color of each element</td>
   <td>yellow</td>
 </tr>

  <tr class="searchbody">
   <th>glyph</th>
   <td>Style of each graphical element (see list below)</td>
   <td>transcript</td>
  </tr>

  <tr class="searchbody">
    <th>height</th>
    <td>Height of each graphical element (pixels)</td>
    <td>10</td>
  </tr>

  <tr class="searchbody">
    <th>key</th>
    <td>Key to the feature.  This is a human-readable description that
    will be printed in the key section of the display</td>
    <td>ESTs aligned via TwinScan 1.2</td>
  </tr>

  <tr class="searchbody">
    <th>label</th>
    <td>Print the feature's name (0=no, 1=yes)</td>
    <td>1</td>
  </tr>

  <tr class="searchbody">
    <th>linewidth</th>
    <td>Width of lines (pixels)</td>
    <td>3</td>
  </tr>

  <tr class="searchbody">
    <th>link</th>
    <td>URL to link to.  This is a Web link in which certain variables
    beginning with the "$" will be replaced with feature attributes.
	Recognized variables are: <b>$name</b> - the name of the
	feature, and <b>$type</b> - the type of the feature (e.g. EST).
    </td>
    <td>link = http://www.your.site/db/get?id=$name;type=$type</td>
  </tr>

  <tr class="searchbody">
     <th>citation</th>
     <td>This is a longer narrative description of the feature intended to
	 identify the author and detailed description of the
	 method. It can be either a text description or a link.</td>
     <td>http://your.site.org/detailed_description.html</td>
  </tr>

  <tr class="searchbody">
     <th>strand_arrow</th>
     <td>Indicate feature strandedness using an arrow (0=no, 1=yes).
	 NB: Strandedness is depicted differently by different
	 glyphs, and in some cases is the default.</td>
     <td>1</td>

   <tr class="searchbody">
     <th>section</th>
     <td>Indicates where in the gbrowse window this type of feature
	 should be placed: "details"=details panel;
	 "overview"=overview panel; "region"=region panel (if there is
	 one for this source);
	 "details+overview"=both panels; "details+overview+region"=all
	 three panels.  <p>
	 "details" is the default.
     </td>
     <td>details+overview</td>
  </tr>

</table>

<P>The bump option is the most important option for controlling the look
of the image. If set to false (the number 0), then the features are
allowed to overlap. If set to true (the number 1), then the features
will move vertically to avoid colliding. If not specified, bump is
turned on if the number of any given type of sequence feature is
greater than 50.</P>

<p>

Unlike the data section, you do not need to put quotes around option
values that contain white space.  In fact, you can continue long
option values across multiple lines by putting extra space in front of
the continuation lines:

<blockquote><pre>
[GenomeAlign]
citation = The pseudoobscura genome was aligned to melanogaster using
   GenomeAlign version 1.0.  High-similarity regions are shown in
   blue, low similarity regions are shown in orange.  The work was
   performed by Joe Postdoc, and is currently in press.
</pre></blockquote>

<p>

Some glyphs also have glyph-specific options.  These are described in
detail below.


<h3>Colors</h3>

<P>Colors are English-language color names or Web-style #RRGGBB colors
(see any book on HTML for an explanation).  The following colors are
recognized:

<p>

<table border="1" class="searchbody"> <TR>
  <TD>white</TD> <TD>coral</TD> <TD>darkslateblue</TD> <TD>green</TD>
  <TD>lightpink</TD> <TD>mediumslateblue</TD> <TD>paleturquoise</TD>
  <TD>sienna</TD></TR> 
<TR>
  <TD>black</TD> <TD>cornflowerblue</TD> <TD>darkslategray</TD>
  <TD>greenyellow</TD> <TD>lightsalmon</TD> <TD>mediumspringgreen</TD>
  <TD>palevioletred</TD> <TD>silver</TD></TR> 
<TR>
  <TD>aliceblue</TD> <TD>cornsilk</TD> <TD>darkturquoise</TD>
  <TD>honeydew</TD> <TD>lightseagreen</TD> <TD>mediumturquoise</TD>
  <TD>papayawhip</TD> <TD>skyblue</TD></TR> 
<TR>
  <TD>antiquewhite</TD> <TD>crimson</TD> <TD>darkviolet</TD>
  <TD>hotpink</TD> <TD>lightskyblue</TD> <TD>mediumvioletred</TD>
  <TD>peachpuff</TD> <TD>slateblue</TD></TR> 
<TR>
  <TD>aqua</TD> <TD>cyan</TD> <TD>deeppink</TD> <TD>indianred</TD>
  <TD>lightslategray</TD> <TD>midnightblue</TD> <TD>peru</TD>
  <TD>slategray</TD></TR> 
<TR>
  <TD>aquamarine</TD> <TD>darkblue</TD> <TD>deepskyblue</TD>
  <TD>indigo</TD> <TD>lightsteelblue</TD> <TD>mintcream</TD>
  <TD>pink</TD> <TD>snow</TD></TR> 
<TR>
  <TD>azure</TD> <TD>darkcyan</TD> <TD>dimgray</TD> <TD>ivory</TD>
  <TD>lightyellow</TD> <TD>mistyrose</TD> <TD>plum</TD>
  <TD>springgreen</TD></TR> 
<TR>
  <TD>beige</TD> <TD>darkgoldenrod</TD> <TD>dodgerblue</TD>
  <TD>khaki</TD> <TD>lime</TD> <TD>moccasin</TD> <TD>powderblue</TD>
  <TD>steelblue</TD></TR> 
<TR>
  <TD>bisque</TD> <TD>darkgray</TD> <TD>firebrick</TD>
  <TD>lavender</TD> <TD>limegreen</TD> <TD>navajowhite</TD>
  <TD>purple</TD> <TD>tan</TD></TR> 
<TR>
  <TD>blanchedalmond</TD> <TD>darkgreen</TD> <TD>floralwhite</TD>
  <TD>lavenderblush</TD> <TD>linen</TD> <TD>navy</TD> <TD>red</TD>
  <TD>teal</TD></TR> 
<TR>
  <TD>blue</TD> <TD>darkkhaki</TD> <TD>forestgreen</TD>
  <TD>lawngreen</TD> <TD>magenta</TD> <TD>oldlace</TD>
  <TD>rosybrown</TD> <TD>thistle</TD></TR> 
<TR>
  <TD>blueviolet</TD> <TD>darkmagenta</TD> <TD>fuchsia</TD>
  <TD>lemonchiffon</TD> <TD>maroon</TD> <TD>olive</TD>
  <TD>royalblue</TD> <TD>tomato</TD></TR> 
<TR>
  <TD>brown</TD> <TD>darkolivegreen</TD> <TD>gainsboro</TD>
  <TD>lightblue</TD> <TD>mediumaquamarine</TD> <TD>olivedrab</TD>
  <TD>saddlebrown</TD> <TD>turquoise</TD></TR> 
<TR>
  <TD>burlywood</TD> <TD>darkorange</TD> <TD>ghostwhite</TD>
  <TD>lightcoral</TD> <TD>mediumblue</TD> <TD>orange</TD>
  <TD>salmon</TD> <TD>violet</TD></TR> 
<TR>
  <TD>cadetblue</TD> <TD>darkorchid</TD> <TD>gold</TD>
  <TD>lightcyan</TD> <TD>mediumorchid</TD> <TD>orangered</TD>
  <TD>sandybrown</TD> <TD>wheat</TD></TR> 
<TR>
  <TD>chartreuse</TD> <TD>darkred</TD> <TD>goldenrod</TD>
  <TD>lightgoldenrodyellow</TD> <TD>mediumpurple</TD> <TD>orchid</TD>
  <TD>seagreen</TD> <TD>whitesmoke</TD></TR> 
<TR>
  <TD>chocolate</TD> <TD>darksalmon</TD> <TD>gray</TD>
  <TD>lightgreen</TD> <TD>mediumseagreen</TD> <TD>palegoldenrod</TD>
  <TD>seashell</TD> <TD>yellow</TD></TR> 
<TR>
  <TD>coral</TD> <TD>darkseagreen</TD> <TD>green</TD>
  <TD>lightgrey</TD> <TD>mediumslateblue</TD> <TD>palegreen</TD>
  <TD>sienna</TD> <TD>yellowgreen</TD></TR>
  </table>

</P>

<h3>Glyphs</h3>

  <P>The ``glyph'' option controls how the features are rendered. The
following glyphs are implemented:</P>
<table border=1>
  <tr class="searchtitle">
  <th> Name</th><th>Description</th>
</tr>
<tr class="searchbody">
  <th>generic</th>
  <td>A filled rectangle.</td>
</tr>
<tr class="searchbody">
  <th>ellipse</th>
  <td>An oval</td>
</tr>
<tr class="searchbody">
<th>arrow</th>
<td>An arrow; can be unidirectional or
    bidirectional.  It is also capable of displaying
    a scale with major and minor tickmarks, and can 
    be oriented horizontally or vertically.
</td>
</tr>
<tr class="searchbody">
<th>segments</th>
<td>A set of filled rectangles connected by solid
    lines. Used for interrupted features, such as 
    gapped alignments and exon groups.
    </td>
</tr>

<tr class="searchbody">
<th>gene</th>
<td>The "gene" glyph is suitable for drawing coding genes. The coding regions
    will be drawn using the specified bgcolor, and the UTRs will be drawn in grey.
    You can change the color of the UTRs by specifying a "utr_color" option.
    For the gene glyph to work properly, the top level feature must be of type
    "mRNA" and the subparts of type "UTR", "five_prime_UTR", "three_prime_UTR",
    or "CDS". See the top of this document for an example.
</td>
</tr>


<tr class="searchbody">
<th>transcript</th>
<td>Similar to segments, but the connecting line is
                      a &quot;hat&quot; shape, and the direction of
                      transcription is indicated by a small arrow.
</td>
</tr>

<tr class="searchbody">
<th>transcript2</th>
<td>Similar to transcript, but the direction of
    transcription is indicated by a terminal segment
    in the shape of an arrow.
</td>
</tr>

<tr class="searchbody">
<th>anchored_arrow</th>
<td>Similar to arrow, but the arrow is drawn in order to take account
    of features whose end-point(s) are unknown, rather than to indicate
    strandedness.
</td>
</tr>

<tr class="searchbody">
<th>primers</th>
<td>Two inward pointing arrows connected by a line. Used for STSs.</td>
</tr>

<tr class="searchbody"> <th>triangle</th> <td>A triangle, used to
represent point features like SNPs, or deletions and insertions.  May
be oriented north, south, east or west.<td> </tr>

<tr class="searchbody"> <th>xyplot</th> <td>A histogram, line plot or column chart,
used for graphic numeric features such as microarray intensity values. To indicate
the value you wish to chart, add a score=XXXX note to the description section:
<blockquote><pre>
    Intensity    expt1    15569..15724    score=192
    Intensity    expt1    17451..17597    score=1071
    </pre></blockquote>
    See <a href="#numeric">Displaying Intensity Plots &amp; Other
    Numeric Data</a> for full details.  <td> </tr>

<tr class="searchbody"> <th>wiggle_xyplot</th> <td>Similar to the
    xyplot glyph, but specialized for displaying very dense
    quantitative data. When you upload a WIG file, this glyph is
    automatically chosen for you.<td> </tr>

<tr class="searchbody"> <th>graded_segments</th> <td>A set of connected segments
    whose colors change intensity according to a score indicated by a "score=XXX" tag.
    The low and high scores are indicated by "min_score" and "max_score" options
    in the configuration stanza, and the basic color is indicated by "bgcolor."</td></tr>

<tr class="searchbody"> <th>wiggle_density</th> <td>Similar to the
    graded_segments glyph, but specialized for displaying very dense
    quantitative data. When you upload a WIG file, this glyph is
    automatically chosen for you as an alternative to wiggle_xyplot.<td> </tr>

<tr class="searchbody"> <th>heat_map</th> <td>A set of connected segments
    whose colors change hue according to a score indicated by a "score=XXX" tag.
    The low and high scores are indicated by "min_score" and "max_score" options
    in the configuration stanza, and the start and ending hues are indicated
    by "start_color" and "end_color." A feature with score equal to min_score will
    be displayed using start_color, while a feature with score equal to max_score
    will be displayed using end_color. Intermediate scores are displayed by blending
    the two hues.
    <td></tr>

<tr class="searchbody">
<th>trace</th>
<td>Reads a SCF sequence file and displays the trace graph. For this glyph to
work, the trace file must be placed on a web-accessible FTP or HTTP server and the
location indicated by a "trace" tag:
    <blockquote><pre>
    Read    ef18222    15569..15724    trace=http://my.host/traces/ef1822.scf
    </pre></blockquote>
</td>
</tr>

</table>

<p>

<a name="glyph_specific">The following glyph-specific options are recognized:</a>

<table border="1">
  <tr class="searchtitle">
    <th>Glyph</th><th>Option</th><th>Description</th></tr>
  </tr>

  <tr class="searchbody">
  <th class="searchtitle">arrow, anchored_arrow</th>
  <td>tick</td>
  <td>Draw major and minor ticks on arrow.
      (0 = no ticks,
      1 = major ticks,
      2 = major &amp; minor ticks)
  </td>

<tr class="searchbody">
 <td>&nbsp;</td>
 <td>parallel</td>
 <td>Whether to draw the arrow
     parallel to the sequence
     or perpendicular to it (1=parallel, 0=antiparallel).
 </td>
</tr>

<tr class="searchbody">
  <td>&nbsp;</td>
  <td>northeast, east</td>
  <td>Force a north or east
      arrowhead.  (The two option names are synonyms.) (0=false, 1=true)
  </td>
</tr>

<tr class="searchbody">
<td>&nbsp;</td>

<td>southwest, west</td>
<td>Force a south or west arrowhead.  (The two option names are
    synonyms.) (0=false, 1=true)</td>
</tr>

<tr class="searchbody">
<td>&nbsp;</td>
<td>double</td>
<td>force doubleheaded arrow (0=false, 1=true)</td>
</tr>

<tr class="searchbody">
<td>&nbsp;</td>
<td>base</td>
<td>Draw a vertical base at the
    non-arrowhead side (0=false, 1=true).
</td>
</tr>

<tr class="searchbody">
<td>&nbsp;</td>
<td>scale</td>
<td>Reset the labels on the arrow
    to reflect an externally 
    established scale.
    </td>
</tr>

<tr class="searchbody">
<th class="searchtitle">gene</th>
<td>utr_color</td>
<td>Color of the UTRs</td>
</tr>

<tr class="searchbody">
<td>&nbsp;</td>
<td>thin_utr</td>
<td>If set to a non-zero value, then UTRs will be drawn as thin boxes</td>
</tr>

<tr class="searchbody">
<td>&nbsp;</td>
<td>decorate_introns</td>
<td>If set to a non-zero value, then introns will be decorated with little
    arrows indicating the direction of the transcript</td>
</tr>

<tr class="searchbody">
<th class="searchtitle">primers</th>
    <td>connect</td>
    <td>Whether to connect the inward-pointing arrowheads by a line
	(0=false, 1=true)</td>
</tr>

<tr class="searchbody">
<th>&nbsp;</th>
    <td>connect_color</td>
    <td>Color of the connecting line</td>
</tr>

<tr class="searchbody">
<th class="searchtitle">triangle</th>
    <td>point</td>
    <td>Is this a point-like feature?  If true, the triantle will be
	drawn at the center of the range, and not scaled to the
	feature width.  (0=false, 1=true)
    </td>
</tr>
<tr class="searchbody">
<th>&nbsp;</th>
    <td>orient</td>
    <td>Orientation of the triangle.  (N=north, S=south, E=east, W=west)
    </td>
</tr>

<tr class="searchbody">
  <th class="searchtitle">xyplot, wiggle_xyplot</th>
  <td>graph_type</td>
  <td>Type of graph
      (histogram, boxes, line, points, linepoints)
  </td>
</tr>

<tr class="searchbody">
  <td>&nbsp;</td>
  <td>min_score</td>
  <td>Minimum score for feature (will be level 0 on graph)  </td>
</tr>

<tr class="searchbody">
  <td>&nbsp;</td>
  <td>max_score</td>
  <td>Maximum score for feature (will be level 0 on graph)  </td>
</tr>

<tr class="searchbody">
  <td>&nbsp;</td>
  <td>scale</td>
  <td>Where to draw the Y axis scale, if any (left, right, both, none)</td>
</tr>

<tr class="searchbody">
  <td>&nbsp;</td>
  <td>point_symbol</td>
  <td>When using points or linepoints graph types, controls the symbol to use for the data points.  One of
    triangle, square, disc, point, or none.</td>
</tr>
    
<tr class="searchbody">
    <td>&nbsp;</td>
  <td>point_radius</td>
  <td>The radius of the symbols, if applicable, in pixels</td>
</tr>

<tr class="searchbody">
    <td>&nbsp;</td>
  <td>bicolor_pivot</td>
  <td>This is a numeric option, which, if specified, causes the
      histogram to be drawn in two colors, one for values above the
      limit specified by the "pos_color" option and one for values below the
      limit specified by the "neg_color" option.</td>
</tr>

<tr class="searchbody">
    <td>&nbsp;</td>
  <td>pos_color</td>
  <td>The color to draw values which are above bicolor_pivot, if
    bicolor_pivot is specified.</td>
</tr>

<tr class="searchbody">
    <td>&nbsp;</td>
  <td>neg_color</td>
  <td>The color to draw values which are below bicolor_pivot, if
    bicolor_pivot is specified.</td>
</tr>

<tr class="searchbody">
<th class="searchtitle">graded_segments, wiggle_density</th>
<td>min_score</td>
<td>Minimum score for the feature (will be drawn as a white segment)</td>
</tr>
    
<tr class="searchbody">
  <td>&nbsp;</td>
  <td>max_score</td>
  <td>Maximum score for the feature (will be drawn as a segment with full intensity bgcolor)</td>
</tr>

<tr class="searchbody">
<th class="searchtitle">heat_map</th>
<td>min_score</td>
<td>Minimum score for the feature (the segment will be drawn with the starting hue)</td>
</tr>
    
<tr class="searchbody">
  <td>&nbsp;</td>
  <td>max_score</td>
  <td>Maximum score for the feature (the segment will be drawn with the ending hue)</td>
</tr>

<tr class="searchbody">
  <td>&nbsp;</td>
  <td>start_color</td>
  <td>Color for segments with the minimum score</td>
</tr>

<tr class="searchbody">
  <td>&nbsp;</td>
  <td>end_color</td>
  <td>Color for segments with the maximum score</td>
</tr>

<tr class="searchbody">
    <td>&nbsp;</td>
  <td>bicolor_pivot</td>
  <td>This is a numeric option, which, if specified, causes the
      histogram to be drawn in two colors, one for values above the
      limit specified by the "pos_color" option and one for values below the
      limit specified by the "neg_color" option.</td>
</tr>

<tr class="searchbody">
    <td>&nbsp;</td>
  <td>pos_color</td>
  <td>The color to draw values which are above bicolor_pivot, if
    bicolor_pivot is specified.</td>
</tr>

<tr class="searchbody">
    <td>&nbsp;</td>
  <td>neg_color</td>
  <td>The color to draw values which are below bicolor_pivot, if
    bicolor_pivot is specified.</td>
</tr>

<tr class="searchbody">
  <th class="searchtitle">trace</th>
  <td>trace</td>
  <td> Specify the trace path or URL to use for this feature.  </td>
</tr>

<tr class="searchbody">
  <td>&nbsp;</td>
  <td>trace_prefix</td>
  <td>String to prepend to each trace path. You may prepend a directory or a
      partial URL.
  </td>
</tr>

<tr class="searchbody">
  <td>&nbsp;</td>
  <td>trace_height</td>
  <td>The height in pixels that the trace will be drawn.</td>
</tr>

<tr class="searchbody">
  <td>&nbsp;</td>
  <td>vertical_spacing</td>
  <td>Vertical distance from the box that shows the physical span of the
      feature to the top of the picture (in pixels).
  </td>
</tr>

<tr class="searchbody">
  <td>&nbsp;</td>
  <td>glyph_delegate</td>
  <td>Glyph to use when zoomed out too far for the trace to be drawn.</td>
</tr>

<tr class="searchbody">
  <td>&nbsp;</td>
  <td>a_color</td>
  <td>Color of the line representing Adenine on the trace.</td>
</tr>

<tr class="searchbody">
  <td>&nbsp;</td>
  <td>c_color</td>
  <td>Color of the line representing Cytosine on the trace.</td>
</tr>

<tr class="searchbody">
  <td>&nbsp;</td>
  <td>g_color</td>
  <td>Color of the line representing Guanine on the trace.</td>
</tr>

<tr class="searchbody">
  <td>&nbsp;</td>
  <td>t_color</td>
  <td>Color of the line representing Thymine on the trace.</td>
</tr>

<tr class="searchbody">
  <td>&nbsp;</td>
  <td>show_border</td>
  <td>Show the black border from around the trace (0=false, 1=true)</td>
</tr>

</table>

<hr>

<H2><a name="numeric">Displaying Intensity Plots &amp; Other Numeric
Data</a></H2>

<p>

For large data sets such as chromosome-wide tiling arrays, please use
<a href="http://genome.ucsc.edu/goldenPath/help/wiggle.html"
target="_blank">Wiggle (WIG)</a> format. However, for smaller data
sets (1-10,000 points), you can use FFF format to achieve a quick and
dirty display.
<p>

Here is a simple template for you to follow. The result is shown on
the right.

<img src="images/help/numeric+data.gif" align="RIGHT">

<table border="1">
  <tr>
    <td>
<pre>
[expression]
glyph = xyplot
graph_type=boxes
fgcolor = black
bgcolor = red
height=100
min_score=0
max_score=110
label=1
key=Expression Level


reference=chr1
expression       liver   10000..12000    score=10
expression       liver   12001..19999    score=50
expression       liver   20000..25000    score=100

expression       kidney   10000..12000    score=80
expression       kidney   12001..19999    score=10
expression       kidney   20000..25000    score=100
</pre>
    </td>
  </tr>
</table>

<p>


The [expression] section says to use the xyplot type of glyph, to set
its type to "boxes" (a column chart), to make the fore and background
colors black and red respectively, to set the height of the chart to
100, and to set the minimum and maximum values for the Y axis to 0 and
110 respectively.  We also add a label and a human-readable track key.

<p>


The data section defines two experiments to show in the track.  Both
experiments use probes whose positions are relative to landmark B0019
(you can of course use chromosome coordinates, or whatever you
choose).  Both experiments are of type "expression", but one is the
"liver" experiment and the other is the "kidney" experiment, as
indicated in the second column.  The third column contains the
coordinates of each assay point, and the fourth column contains the
score=XXX attribute, where XXX is the intensity value.

<p>

See the <a href="#glyph_specific">chart in the previous section</a>
for the xyplot glyph options.

<hr>

<H2><a name="remote">Importing Custom Tracks via the Internet</a></H2>

<p>

GBrowse can display annotation files that are physically located on
internet-connected sites. Use the "Import a track" section to paste in
the URL of an annotation file. The following URL types are allowed:

<p>

<dl>
  <dt><b>The URL of a BED, GFF3, or FFF file.</b>
  <dd>If a BED, GFF3 or FFF file is located on an internet-accessible
      FTP or web server, paste its URL into the "remote URL" text
      field. It will be mirrored to the GBrowse server and
      displayed. When you update the file, the updated version will be
      mirrored automatically.

  <dt><b>The URL of a BigWig (.bw) file</b> <dd>This is a format that
  allows you to import UCSC Wiggle tracks into the browser in a fast
  and efficient manner.  If you haven't already done so, obtain the
  BigWig tools from the <a
  href="http://genome.ucsc.edu/goldenPath/help/bigWig.html">UCSC
  Genome Browser</a> and run the wigToBigWig utility as described
  there.  Name the file with a ".bw" extension (important!) and place
  it on an FTP or Web server that can be accessed via this
  browser. Finally paste the full URL of the BigWig file into the
  "Import a track URL" field and press "upload". Note: To create the
  BigWig file you will need a "chrom.sizes" file corresponding to the
  chromosome lengths of the genome. To obtain a suitable file from the
  browser, select <i>Download Chrom Sizes</i> from the "File" menu in
  the upper left corner of the genome browser page.

  <dt><b>The URL of a sorted, indexed BAM file</b>
  <dd>Sort a BAM sequence alignment file and index it with
      SamTools. Place the BAM file and its associated .bai index file on
      a web or FTP-accessible server and paste its URL into the "Import a
      track URL" field.
      The information in the file
      will be retrieved as needed in a network-efficient manner. It is up to
      you to ensure that the chromosome coordinates of the BAM file match the
      build of the genome used by this instance of the browser, or nothing will
      show up (you'll get blank tracks). Please be aware that the URL must end
      with ".bam" for GBrowse to recognize that it is a BAM file.

  <dt><b>A GBrowse or DAS URL</b>
  <dd>You can share tracks from one GBrowse
      server to another by clicking on the <img
      src="images/buttons/share.png"> icon. This will give you a URL
      that you can paste into the URL import field. Alternatively, you
      can view a large variety of annotations using the distributed
      annotation system (DAS) protocol. The <a
      href="http://www.dasregistry.org/">DAS registry</a> will help you
      locate DAS servers with useful genomic data. Then cut and paste the
      DAS server URL into the remote track URL box as before.
</dl>
  <hr>

$Id$

</body>
</html>
