<h2>gbrowse_karyotype - image generator</h2>

<h3>Using the URL query string to draw chromosomes and map sequence features</h3>

<h2>SYNOPSIS</h2>
<pre>  &lt;img src="http://mckay.cshl.edu/cgi-bin/gbrowse_karyotype?source=human"&gt;<br>
<br>
  <i>Will generate this picture:</i><br>
 
<img alt="karyotype image"
 src="http://mckay.cshl.edu/cgi-bin/gbrowse_karyotype?source=human">
<br>
<br>
  &lt;a href="http://mckay.cshl.edu/cgi-bin/gbrowse_karyotype?list=sources"&gt;list&lt;/a&gt;<br>
<br>
  <i>Will return this document:</i><br>
<iframe
 src="http://mckay.cshl.edu/cgi-bin/gbrowse_karyotype?list=sources"
 width="500">
iframes not supported; click <a href="http://mckay.cshl.edu/cgi-bin/gbrowse_karyotype?list=sources">here</a>
</iframe>
<br>
  &lt;a href="http://mckay.cshl.edu/cgi-bin/gbrowse_karyotype/example?list=types"&gt;types&lt;/a&gt;<br>
<iframe src="http://mckay.cshl.edu/cgi-bin/gbrowse_karyotype/example?list=types"
 width="500">
 iframes not supported; click <a href="http://mckay.cshl.edu/cgi-bin/gbrowse_karyotype?list=types">here</a>
</iframe>
<br>
</pre>
<h2>DESCRIPTION</h2>
<p>
This CGI script is an interface to the Generic Genome Browser for the
purpose of drawing a stylized karyotype for a particular genome. It
can be used as the destination of an &lt;img&gt; tag like this:
</p>
<blockquote>
  <pre>&lt;img src="http://mckay.cshl.edu/cgi-bin/gbrowse_karyotype?source=mouse"&gt;<br>
  </pre>
</blockquote>
<p>or as a standalone image-mapped HTML.<br>
</p>
<p>The script can also be used to superimpose one or more external
features onto the display, for example for the purpose of displaying
genes in the context of their chromosomal position.
</p>
<h2>CGI arguments</h2>
<p>
The script recognizes the following CGI arguments, which can be passed
either as GET or POST argument=value pairs. Argument pairs must be
separated by semicolons (preferred) or by ampersands.&nbsp; Most
options have one-letter aliases that can be used to reduce URL
lengths.
</p>
<table style="background-color: rgb(255, 255, 204);" border="1">
  <tbody>
    <tr>
      <th>Argument</th>
      <th>Alias</th>
      <th>Description</th>
    </tr>
    <tr>
      <td>type</td>
      <td>t</td>
      <td>built-in tracks to include in image</td>
    </tr>
    <tr>
      <td>rows<br>
      </td>
      <td>r<br>
      </td>
      <td>number of rows of chromosomes to draw<br>
      </td>
    </tr>
    <tr>
      <td>options</td>
      <td>o</td>
      <td>list of track options (compact, labeled, etc)</td>
    </tr>
    <tr>
      <td>chromosome<br>
      </td>
      <td>c<br>
      </td>
      <td>which chromosome(s) to draw<br>
      </td>
    </tr>
    <tr>
      <td>add</td>
      <td>a</td>
      <td>added feature(s) to superimpose on the image</td>
    </tr>
    <tr>
      <td>style</td>
      <td>s</td>
      <td>stylesheet for additional features</td>
    </tr>
    <tr>
      <td>embed</td>
      <td><br>
      </td>
      <td>generate full HTML for image and imagemap for use in an
embedded frame</td>
    </tr>
    <tr>
      <td>cwidth<br>
      </td>
      <td>w<br>
      </td>
      <td>chromsome width<br>
      </td>
    </tr>
    <tr>
      <td>cheight<br>
      </td>
      <td>h<br>
      </td>
      <td>maximum chromosome height<br>
      </td>
    </tr>
    <tr>
      <td>bandlabels<br>
      </td>
      <td>b<br>
      </td>
      <td>label chromosome bands<br>
      </td>
    </tr>
    <tr>
      <td>list</td>
      <td>&nbsp;l</td>
      <td>get certain types of configuration information</td>
    </tr>
    <tr>
      <td>source</td>
      <td>&nbsp;src</td>
      <td>database name</td>
    </tr>
  </tbody>
</table>
<p>The arguments are explained in more detail here<br>
</p>
<dl>
  <dd><br>
  </dd>
  <dt><b>type</b> (Alias: <b>t</b>) </dt>
  <dd>This argument lists the feature types to display. The value of
this argument is a list of track names separated by spaces ("+"
characters when URL-escaped). For example:
    <p> </p>
    <pre>&lt;img src="http://mckay.cshl.edu/cgi-bin/gbrowse_karyotype?src=rat;<br>
                   type=gene"&gt;<br>
      </pre>
Multiple <i>type=</i> arguments will be combined to form a single
space-delimited list. The alias "t" can be used to shorten the length
of the URL.
    <p> If the track name has a space in it, put quotes around the
name: </p>
    <pre>            type="microbe tRNA"+NG+WABA+CG+ESTB<br>
      </pre>
    <p> </p>
  </dd>
  <dt><b>cwidth</b> (Alias: <b>w</b>) </dt>
  <dd>Width of the chromosomes, in pixels.&nbsp; The default value can
be specified in the source configuration file.</dd>
  <dt><br>
  </dt>
  <dt><b>cheight</b> (Alias: <span style="font-weight: bold;">h</span>)
  </dt>
  <dd>Height of the largest chromosome, in pixels.&nbsp; The default
value can be specified in the source configuration file.</dd>
  <dt><br>
  </dt>
  <dt><b>rows</b> (Alias: <span style="font-weight: bold;">r</span><b></b>)
  </dt>
  <dd>The number or rows to arrange the chromosomes in.&nbsp; The
default behavior is to automatically adjust the number of rows to</dd>
  <dd>keep the image less than 1024 pixels in width.</dd>
  <dt><br>
  </dt>
  <dt><b>chromome</b> (Alias: <span style="font-weight: bold;">c</span><b></b>)
  </dt>
  <dd>Number and order of chromosomes to draw.&nbsp; The default is to
draw all chromosomes.&nbsp; This can also be specified in the source
configuration file.&nbsp; Multiple chromosomes can be specified in a
space-delimited list ("+" characters when URL-escaped).&nbsp; For
example c=1+2+3+4 will draw chromosomes 1-4.<br>
  </dd>
</dl>
<b>options</b> (Alias: <b>o</b>)
<dl>
  <dd>A space-delimited list ("+" characters when URL-escaped) of
mnemonic/option pairs describing how features should be formatted.
Options are integers from 0 to 3, where 0=auto, 1=compact, 2=expanded,
3=expanded and labeled. For example, to specify that the tRNA and NG
tracks should always be expanded and labeled, but that the WABA track
should be compact, use:
    <p> </p>
    <pre>      options=tRNA+3+NG+3+WABA+1<br>
      </pre>
    <p> The alias "o" can be used to shorten the length of the URL. </p>
    <p> </p>
  </dd>
  <dt><b>add</b> (Alias: <b>a</b>) </dt>
  <dd>Superimpose one or more additional features on top of the view.
Features are specified as space ("+") delimited lists in the following
format:
    <p> </p>
    <pre>      add=Chromosome+Type+Name+start..end,start..end,start..end<br>
      </pre>
"Type" is a descriptive type that will
be printed in the image caption, "Name" is a name for the feature, and
start..end is a comma-delimited list of ranges
for discontinuous feature. Names that contain white space must be
quoted, for example "BLAST hit". Note that this all has to be
URL-escaped, so an additional feature named "Your Sequence", type
"Blast Hit", that is located on chromosome III in a gapped range
between 20000 and 22000, will be formatted as:
    <p> </p>
    <pre>      add=III+%22Blast%20Hit%22+%22Your%20Sequence%22+20000..21000,21550..22000<br>
      </pre>
    <p> One or both of the type and name can be omitted. If omitted,
type will default to "Your Features" and the name will default to
"Feature XX" where XX is an integer. This allows for a very simple
feature line: </p>
    <pre>      add=III+20000..21000,21550..22000<br>
      </pre>
    <p> Multiple <i>add=</i> arguments are allowed. The alias "a" can
be used to shorten the length of the URL. </p>
    <p> </p>
  </dd>
  <dt><b>style</b> </dt>
  <dd>The style argument can be used to control the rendering of
additional features added with "add". It is a flattened version of the
style configuration sections described in <a
 href="/cgi-bin/gbrowse?help=annotation">this
document</a> For example, if you have added a "Blast Hit" annotation,
then you can tell the renderer to use a red arrow for this glyph in
this way: style=%22Blast%20Hit%22+glyph=arrow+fgcolor=red
    <p> </p>
  </dd>
  <dt><b>embed</b> </dt>
  <dd>Generate image and a corresponding HTML imagemap in a form
suitable for embedding into a frame.&nbsp; This option is off by
default.
    <p> </p>
  </dd>
  <dt><b>list</b> </dt>
  <dd>If this argument is present, it will cause the script to dump out
various types of information in plain text form. Currently the two
values for this argument are <i>sources</i>, to dump out the list of
data sources, and <i>types</i>, to dump out the list of configured
types. For <i>list=sources</i>, the script will return a simple text
list of the data source names. For <i>list=types</i>, the script will
return a three-column tab-delimited list giving the track names and
feature types corresponding to the currently-selected data source. The
format is as follows:
    <blockquote>
      <pre>      Mnemonic &lt;tab&gt; Full description of feature &lt;tab&gt; [default]<br>
      </pre>
    </blockquote>
    <p> The third column contains the word "default" if the track will
be shown by default when no <i>type</i> argument is provided. </p>
    <p> </p>
  </dd>
  <dt><b>source</b> </dt>
  <dd>This argument specifies the database for the images. The list of
sources can be found using <i>list=sources</i>.
    <p> </p>
  </dd>
</dl>
<p>Putting it all together, here's a working (very long) URL:
</p>
<pre><a
 href="http://mckay.cshl.edu/cgi-bin/gbrowse_karyotype?src=mouse;b=1;c=1+2+3+4;add=1+gene+gene1+10000001..10006000;add=1+gene+gene2+20000005..20000500;s=gene+glyph=triangle+bgcolor=red+fgcolor=red+orient=N+width=6+point=1+bump=1">http://mckay.cshl.edu/cgi-bin/gbrowse_karyotype?src=mouse;b=1;c=1+2+3+4;add=1+gene+gene1+10000001..10006000;add=1+gene+gene2+20000005..20000500;s=gene+glyph=triangle+bgcolor=red+fgcolor=red+orient=N+width=6+point=1+bump=1</a>

</pre>
<p>
If you wish to associate the image with an imagemap so that clicking
on a feature takes the user to the destination configured in the
gbrowse config file, you may do so by placing the URL in an
&lt;iframe&gt; section and using the <b>embed=1</b> flag:
</p>
<pre>&lt;iframe src="http://mckay.cshl.edu/cgi-bin/gbrowse_karyotype?src=mouse;embed=1" width="100%" height="250"&gt;<br>
   &lt;img src="http://mckay.cshl.edu/cgi-bin/gbrowse_karyotype?src=mouse;embed=1"/&gt;<br>
&lt;/iframe&gt;<br>
</pre>
<p>
Placing an &lt;img&gt; tag inside the &lt;iframe&gt; tag arranges for
older browsers that don't know about iframes to display the static
image instead. You may need to adjust the width and height attributes
in order to avoid browsers placing scrollbars around the frame.
</p>
<h2>KNOWN BUGS</h2>
<p>Please report them to the author.<br>
</p>
<h2>AUTHOR</h2>
<p>Sheldon McKay&nbsp;&nbsp; <a href="mailto:mckays%5C@cshl.edu">mckays@cshl.edu</a>
</p>
<p>
This documentation is based on part on gbrowse_img, written by&nbsp;
<a href="mailto:lstein%5C@cshl.edu">Lincoln Stein</a>

</p>
<p>Copyright (c) 2006 Cold Spring Harbor Laboratory
</p>
<p>This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.
</p>
<p>For additional help, see <a href="http://www.gmod.org">The GMOD
Project</a> pages.</p>

