<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
<html> <head>
<title>Generic Genome Browser</title>
<link rel="stylesheet" href="tutorial/tutorial.css">
</head>

<body>
<h1>Generic Genome Browser</h1>

<a href="http://www.gmod.org/"><img src="gmod_cog.jpeg" border="0" align="RIGHT"></a>

<p>

Welcome to the Generic Genome Browser!
For future reference, here is
where you've installed its various components:

<dl>
  <dt>GBrowse documentation, stylesheets and in-memory databases.
  <dd><b>$HTDOCS/gbrowse</b> (also available for browsing at
      <a href="http://localhost/gbrowse">http://localhost/gbrowse</a>)
      <p>
  <dt>GBrowse configuration files
  <dd><b>$CONF/gbrowse.conf</b>
      <p>
  <dt>GBrowse CGI (web) scripts
  <dd><b>$CGIBIN/gbrowse</b>, <i>$CGIBIN/gbrowse_img</i>,<i>$CGIBIN/gbrowse_details</i>
      <p>
  <dt>Online tutorial
  <dd><b>$HTDOCS/gbrowse/tutorial</b> (available for browsing at <a href="http://localhost/gbrowse/tutorial/tutorial.html">http://localhost/gbrowse/tutorial/tutorial.html</a>).
      <p>
  <dt>Reference manual
  <dd><b>$HTDOCS/gbrowse/CONFIGURE_HOWTO.txt</b> (available for browsing at <a href="http://localhost/gbrowse/CONFIGURE_HOWTO.txt">http://localhost/gbrowse/CONFIGURE_HOWTO.txt</a>).
      <p>
  <dt>Sample data
  <dd><b>$HTDOCS/gbrowse/sample_data.txt</b>
      <p>
  <dt>Contributed modules and utilities
  <dd><b>$HTDOCS/gbrowse/contrib</b>
      <p>
</dl>

<p>

<h2>Example Database</h2>

<p>

An example in-memory database from yeast chromosome 1 has been
installed for you.  Try it at this URL:

<blockquote>
<a href="/cgi-bin/gbrowse?source=yeast_chr1">http://localhost/cgi-bin/gbrowse?source=yeast_chr1</a>
</blockquote>


<p>

To load larger data sets it is suggested that you create a MySQL
database.  Sample data from yeast is located in the gbrowse
distribution in the directory <i>sample_data</i>.  Instructions for
loading this data is in <a href="INSTALL">INSTALL</a>.  You will most
likely want to use the script <i>bp_bulk_load_gff.pl</i> which should
have been installed when you installed BioPerl.

<p>

The full yeast genome is also available for download from Sourceforge.
Get <a
href="http://prdownloads.sourceforge.net/gmod/yeast.gff.gz">yeast
features</a> and <a
href="http://prdownloads.sourceforge.net/gmod/yeast.fasta.gz">yeast
DNA</a> from <a
href="http://sourceforge.net/project/showfiles.php?group_id=27707">the
gmod downloads page</a>.  You'll also find Worm, Fly and Human genome
annotations here, although not necessarily the most up to date ones.
Note that you will need to uncompress the data files prior to running
to load script.  See the <a href="contrib/conf_files">sample
configuration files</a> directory for starter config files to get you
up and running.

<p>

Once you've loaded the yeast data (and adjusted the config file as
described in the <a href="INSTALL">INSTALL</a> document), you can
browse the yeast genome by following the link <a
href="/cgi-bin/gbrowse?source=yeast">http://localhost/cgi-bin/gbrowse?source=yeast</a>.

<h2>Learning More</h2>

<p>

To find out more about Gbrowse, try the:

<ul>
  <li><a href="tutorial/tutorial.html">GBrowse Tutorial</a>
  <li><a href="contrib/conf_files">Sample Configuration Files</a>
  <li><a href="CONFIGURE_HOWTO.txt">GBROWSE CONFIGURE HOWTO</a>
</ul>


<hr>
<address>Lincoln D. Stein, lstein@cshl.org<br>
<a href="/">Cold Spring Harbor Laboratory</a></address>
<!-- hhmts start -->
Last modified: Sun Jun 29 14:17:07 PDT 2003
<!-- hhmts end -->
</body> </html>
