<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
<html> <head>
<title>Generic Genome Browser</title>
<link rel="stylesheet" href="tutorial/tutorial.css">
</head>

<body>
<h1>Generic Genome Browser, v2</h1>

<a href="http://www.gmod.org/"><img src="gmod_cog.jpeg" border="0" align="RIGHT"></a>

<p>

Welcome to the Generic Genome Browser!

<h2>Example Database</h2>

<p>

An example in-memory database from yeast chromosome 1 has been
installed for you.  Try it at this URL:

<blockquote>
<a href="/cgi-bin/gb2/gbrowse/yeast_chr1">/cgi-bin/gb2/gbrowse/yeast_chr1</a>
</blockquote>


<h2>Directory Paths</h2>

<p>

For future reference, here is where you've installed GBrowse's various
components:

<dl>
  <dt>GBrowse documentation, stylesheets and in-memory databases.
  <dd><b>$HTDOCS</b> (also available for browsing at
      <a href="/gbrowse">http://localhost/gbrowse</a>)
      <p>
  <dt>GBrowse configuration files
  <dd><b>$CONF/gbrowse.conf</b>
      <p>
  <dt>GBrowse CGI (web) scripts
  <dd><ul>
          <li><b>$CGIBIN/gbrowse</b>
          <li><b>$CGIBIN/gbrowse_img</b>
          <li><b>$CGIBIN/gbrowse_details</b>
          <li><b>$CGIBIN/gbrowse_est</b>
          <li><b>$CGIBIN/gbrowse_syn</b>
          <li><b>$CGIBIN/gbrowse_moby</b>
      </ul>
      <p>
  <dt>Online tutorial
  <dd><b>$HTDOCS/tutorial</b> (available for browsing at <a href="tutorial/tutorial.html">http://localhost/gbrowse/tutorial/tutorial.html</a>).
      <p>
  <dt>Reference manual
  <dd><b>$HTDOCS/docs/pod/INSTALL.html</b> (available for browsing at <a href="docs/pod/INSTALL.html">http://localhost/gbrowse/docs/pod/INSTALL.html</a>).
    <p>
  <dd><b>$HTDOCS/docs/pod/CONFIGURE_HOWTO.html</b> (available for browsing at <a href="docs/pod/CONFIGURE_HOWTO.html">http://localhost/gbrowse/docs/pod/CONFIGURE_HOWTO.html</a>).
    <p>
  <dd><b>$HTDOCS/docs/pod/MAKE_IMAGES_HOWTO.html</b> gives directions for creating stand-alone pictures with the gbrowse_img script (at <a href="docs/pod/MAKE_IMAGES_HOWTO.html">http://localhost/gbrowse/docs/pod/MAKE_IMAGES_HOWTO.html</a>).
    <p>
  <dd><b>$HTDOCS/docs/pod/README-gff-files.html</b> describes methods for obtaining GBrowse compatible GFF files from WormBase, FlyBase, SGD and NCBI/human (at <a href="docs/pod/README-gff-files.html">http://localhost/gbrowse/docs/pod/README-gff-files.html</a>).
    <p>
  <dd><b>$HTDOCS/docs/pod/PLUGINS_HOWTO.html</b> gives directions for creating plugins (at <a href="docs/pod/PLUGINS_HOWTO.html">http://localhost/gbrowse/docs/pod/PLUGINS_HOWTO.html</a>).
    <p>
  <dd><b>$HTDOCS/docs/pod/README-chado.html</b> gives notes for using GBrowse with chado (at <a href="docs/pod/README-chado.html">http://localhost/gbrowse/docs/pod/README-chado.html</a>).
    <p>
  <dd><b>$HTDOCS/docs/pod/DAS_HOWTO.html</b> gives notes for using GBrowse as a Distributed Annotation System (DAS) server and client (at <a href="docs/pod/DAS_HOWTO.html">http://localhost/gbrowse/docs/pod/DAS_HOWTO.html</a>).
    <p>
  <dd><b>$HTDOCS/docs/pod/ORACLE_AND_POSTGRESQL.html</b> gives notes for running GBrowse on top of Oracle and PostgreSQL database servers (at <a href="docs/pod/ORACLE_AND_POSTGRESQL.html">http://localhost/gbrowse/docs/pod/ORACLE_AND_POSTGRESQL.html</a>).
      <p>
  <dd><b>$HTDOCS/docs/pod/BIOSQL_ADAPTER_HOWTO.html</b> has instructions for running GBrowse on top of a BioSQL database (at <a href="docs/pod/BIOSQL_ADAPTER_HOWTO.html">http://localhost/gbrowse/docs/pod/BIOSQL_ADAPTER_HOWTO.html</a>).
      <p>
  <dd><b>$HTDOCS/docs/pod/GENBANK_HOWTO.html</b> has notes for loading and browsing GenBank formated files (at <a href="docs/pod/GENBANK_HOWTO.html">http://localhost/gbrowse/docs/pod/GENBANK_HOWTO.html</a>).
      <p>
  <dd><b>$HTDOCS/docs/pod/INSTALL.MacOSX.html</b> has notes for installing GBrowse on MacOSX (at <a href="docs/pod/INSTALL.MacOSX.html">http://localhost/gbrowse/docs/pod/INSTALL.MacOSX.html</a>).
      <p>
  <dt>Sample data
  <dd><b>$HTDOCS/sample_data.txt</b>
      <p>
  <dt>Contributed modules and utilities
  <dd><b>$HTDOCS/contrib</b>
      <p>
</dl>

<p>

<h2>More Data Sets</h2>

To load larger data sets it is suggested that you create a MySQL
database.  Sample data from yeast is located in the gbrowse
distribution in the directory <i>sample_data</i>.  Instructions for
loading this data is in <a href="docs/pod/INSTALL.html">INSTALL</a>.  You will most
likely want to use the script <i>bp_bulk_load_gff.pl</i> which should
have been installed when you installed BioPerl.

<p>

The full yeast genome is also available for download from Sourceforge.
Get <a
href="http://prdownloads.sourceforge.net/gmod/yeast.gff.gz">yeast
features</a> and <a
href="http://prdownloads.sourceforge.net/gmod/yeast.fasta.gz">yeast
DNA</a> from <a
href="http://sourceforge.net/project/showfiles.php?group_id=27707">the
gmod downloads page</a>.  You'll also find Worm, Fly and Human genome
annotations here, although not necessarily the most up to date ones.
Note that you will need to uncompress the data files prior to running
to load script.  See the <a href="contrib/conf_files">sample
configuration files</a> directory for starter config files to get you
up and running.

<p>

Once you've loaded the yeast data (and adjusted the config file as
described in the <a href="docs/pod/INSTALL.html">INSTALL</a> document), you can
browse the yeast genome by following the link <a
href="/cgi-bin/gbrowse/yeast">http://localhost/cgi-bin/gbrowse/yeast</a>.

<h2>Learning More</h2>

<p>

To find out more about Gbrowse, try the:

<ul>
  <li><a href="tutorial/tutorial.html">GBrowse Tutorial</a>
  <li><a href="contrib/conf_files">Sample Configuration Files</a>
  <li><a href="docs/pod/CONFIGURE_HOWTO.html">GBROWSE CONFIGURE HOWTO</a>
</ul>


<hr>
<address>Lincoln D. Stein, lstein@cshl.org<br>
<a href="http://www.cshl.org/">Cold Spring Harbor Laboratory</a></address>
<!-- hhmts start -->
Last modified: Wed Nov  5 12:07:54 EST 2008
<!-- hhmts end -->
</body> </html>
